`swift build` fails on FreeBSD

## Summary

`swift build` fails on FreeBSD.

- FreeBSD 15.0-RELEASE-p1 x86_64
- Swift version 6.3-dev (LLVM b58b2a34d509492, Swift cf535d8b998d09b) ([Swift on FreeBSD Preview](https://forums.swift.org/t/swift-on-freebsd-preview/83064))
  - Target: x86_64-unknown-freebsd14.3
  - Build config: +assertions
- swift-system main branch 7e474cb4bdde782a94c3568d961265b5c6bf439d

## Error logs

Initial state:

```console
$ swift build --build-tests
...
/home/kebo/swift-system/Sources/System/Internals/Constants.swift:697:33: error: cannot convert return expression of type 'Int32' to return type 'mode_t' (aka 'UInt16')
695 | #if SYSTEM_PACKAGE_DARWIN || os(FreeBSD)
696 | @_alwaysEmitIntoClient
697 | internal var _S_IFWHT: mode_t { S_IFWHT }
    |                                 `- error: cannot convert return expression of type 'Int32' to return type 'mode_t' (aka 'UInt16')
698 | #endif
699 | 

[#ImplementationOnlyDeprecated]: <https://docs.swift.org/compiler/documentation/diagnostics/implementation-only-deprecated>
/home/kebo/swift-system/Sources/System/Internals/Constants.swift:697:33: error: cannot convert return expression of type 'Int32' to return type 'mode_t' (aka 'UInt16')
695 | #if SYSTEM_PACKAGE_DARWIN || os(FreeBSD)
696 | @_alwaysEmitIntoClient
697 | internal var _S_IFWHT: mode_t { S_IFWHT }
    |                                 `- error: cannot convert return expression of type 'Int32' to return type 'mode_t' (aka 'UInt16')
698 | #endif
699 |
...
```

After fixing the error above:

```console
$ swift build --build-tests
...
/home/kebo/swift-system/Tests/SystemTests/StatTests.swift:372:19: error: type 'FileFlags' has no member 'tracked'
370 |       let userSettableFlags: FileFlags = [
371 |         .noDump, .userImmutable, .userAppend,
372 |         .opaque, .tracked, .hidden,
    |                   `- error: type 'FileFlags' has no member 'tracked'
373 |         .userNoUnlink,
374 |         .offline,
...
```

After fixing the error above:

```console
$ swift build --build-tests
...
/home/kebo/swift-system/Tests/SystemTests/StatTests.swift:387:48: error: cannot convert value of type 'CInterop.FileFlags' (aka 'UInt32') to expected argument type 'UInt'
385 | 
386 |       flags.insert(userSettableFlags)
387 |       try #require(fchflags(fd.rawValue, flags.rawValue) == 0, "\(Errno.current)")
    |                                                `- error: cannot convert value of type 'CInterop.FileFlags' (aka 'UInt32') to expected argument type 'UInt'
388 | 
389 |       stat = try fd.stat()
...
```

After fixing the error above:

```console
$ swift build --build-tests
...
/home/kebo/swift-system/Tests/SystemTests/StatTests.swift:393:48: error: cannot convert value of type 'CInterop.FileFlags' (aka 'UInt32') to expected argument type 'UInt'
391 | 
392 |       flags.remove(userSettableFlags)
393 |       try #require(fchflags(fd.rawValue, flags.rawValue) == 0, "\(Errno.current)")
    |                                                `- error: cannot convert value of type 'CInterop.FileFlags' (aka 'UInt32') to expected argument type 'UInt'
394 | 
395 |       stat = try fd.stat()
...
```

--- a/Sources/System/Internals/Constants.swift
+++ b/Sources/System/Internals/Constants.swift
@@ -692,9 +692,12 @@ internal var _S_IFLNK: mode_t { S_IFLNK }
 @_alwaysEmitIntoClient
 internal var _S_IFSOCK: mode_t { S_IFSOCK }
 
-#if SYSTEM_PACKAGE_DARWIN || os(FreeBSD)
+#if SYSTEM_PACKAGE_DARWIN
 @_alwaysEmitIntoClient
 internal var _S_IFWHT: mode_t { S_IFWHT }
+#elseif os(FreeBSD)
+@_alwaysEmitIntoClient
+internal var _S_IFWHT: mode_t { .init(S_IFWHT) }
 #endif
 
 // MARK: - stat/chflags File Flags

--- a/Sources/System/Internals/Constants.swift
+++ b/Sources/System/Internals/Constants.swift
@@ -692,10 +692,8 @@ internal var _S_IFLNK: mode_t { S_IFLNK }
 @_alwaysEmitIntoClient
 internal var _S_IFSOCK: mode_t { S_IFSOCK }
 
-#if SYSTEM_PACKAGE_DARWIN
-@_alwaysEmitIntoClient
-internal var _S_IFWHT: mode_t { S_IFWHT }
-#elseif os(FreeBSD)
+#if SYSTEM_PACKAGE_DARWIN || os(FreeBSD)
+// `S_IFWHT` is `Int32` on FreeBSD.
 @_alwaysEmitIntoClient
 internal var _S_IFWHT: mode_t { .init(S_IFWHT) }
 #endif

--- a/Tests/SystemTests/StatTests.swift
+++ b/Tests/SystemTests/StatTests.swift
@@ -381,13 +381,15 @@ private struct StatTests {
       #endif
 
       flags.insert(userSettableFlags)
-      try #require(fchflags(fd.rawValue, flags.rawValue) == 0, "\(Errno.current)")
+      // On FreeBSD, the second argument of `fchflags` requires `UInt` instead of`UInt32`.
+      try #require(fchflags(fd.rawValue, .init(flags.rawValue)) == 0, "\(Errno.current)")
 
       stat = try fd.stat()
       #expect(stat.flags == flags)
 
       flags.remove(userSettableFlags)
-      try #require(fchflags(fd.rawValue, flags.rawValue) == 0, "\(Errno.current)")
+      // On FreeBSD, the second argument of `fchflags` requires `UInt` instead of`UInt32`.
+      try #require(fchflags(fd.rawValue, .init(flags.rawValue)) == 0, "\(Errno.current)")
 
       stat = try fd.stat()
       #expect(stat.flags == flags)
@@ -413,12 +415,5 @@ private func > (lhs: timespec, rhs: timespec) -> Bool {
 private func == (lhs: timespec, rhs: timespec) -> Bool {
   lhs.tv_sec == rhs.tv_sec && lhs.tv_nsec == rhs.tv_nsec
 }
-
-#if os(FreeBSD)
-// Helper functions for FreeBSD
-private func fchflags(_ fd: CInt, _ flags: CInterop.FileFlags) -> CInt {
-  fchflags(fd, UInt(flags))
-}
-#endif
 
 #endif

--- a/Tests/SystemTests/StatTests.swift
+++ b/Tests/SystemTests/StatTests.swift
@@ -118,11 +118,7 @@ private struct StatTests {
       #expect(targetStat.type == .regular)
       #expect(symlinkStat.type == .symbolicLink)
       #expect(symlinkStat.size < targetStat.size)
-      #if os(FreeBSD)
       #expect(symlinkStat.sizeAllocated <= targetStat.sizeAllocated)
-      #else
-      #expect(symlinkStat.sizeAllocated < targetStat.sizeAllocated)
-      #endif
 
       // Set each .st_atim back to its original value for comparison
 

--- a/Tests/SystemTests/StatTests.swift
+++ b/Tests/SystemTests/StatTests.swift
@@ -385,13 +385,13 @@ private struct StatTests {
       #endif
 
       flags.insert(userSettableFlags)
-      try #require(fchflags(fd.rawValue, UInt(flags.rawValue)) == 0, "\(Errno.current)")
+      try #require(fchflags(fd.rawValue, flags.rawValue) == 0, "\(Errno.current)")
 
       stat = try fd.stat()
       #expect(stat.flags == flags)
 
       flags.remove(userSettableFlags)
-      try #require(fchflags(fd.rawValue, UInt(flags.rawValue)) == 0, "\(Errno.current)")
+      try #require(fchflags(fd.rawValue, flags.rawValue) == 0, "\(Errno.current)")
 
       stat = try fd.stat()
       #expect(stat.flags == flags)
@@ -417,5 +417,12 @@ private func > (lhs: timespec, rhs: timespec) -> Bool {
 private func == (lhs: timespec, rhs: timespec) -> Bool {
   lhs.tv_sec == rhs.tv_sec && lhs.tv_nsec == rhs.tv_nsec
 }
+
+#if os(FreeBSD)
+// Helper functions for FreeBSD
+private func fchflags(_ fd: CInt, _ flags: CInterop.FileFlags) -> CInt {
+  fchflags(fd, UInt(flags))
+}
+#endif
 
 #endif

--- a/Tests/SystemTests/StatTests.swift
+++ b/Tests/SystemTests/StatTests.swift
@@ -118,7 +118,11 @@ private struct StatTests {
       #expect(targetStat.type == .regular)
       #expect(symlinkStat.type == .symbolicLink)
       #expect(symlinkStat.size < targetStat.size)
+      #if os(FreeBSD)
+      #expect(symlinkStat.sizeAllocated <= targetStat.sizeAllocated)
+      #else
       #expect(symlinkStat.sizeAllocated < targetStat.sizeAllocated)
+      #endif
 
       // Set each .st_atim back to its original value for comparison
 
@@ -368,14 +372,11 @@ private struct StatTests {
       ]
       #elseif os(FreeBSD)
       let userSettableFlags: FileFlags = [
-        .noDump, .userImmutable, .userAppend,
-        .opaque, .tracked, .hidden,
-        .userNoUnlink,
-        .offline,
-        .readOnly,
-        .reparse,
-        .sparse,
-        .system
+        .noDump, .hidden, .offline,
+        .readOnly, .sparse, .system
+        // The following flags throw EPERM on ZFS.
+        // .userImmutable, .userAppend, .opaque,
+        // .userNoUnlink, .reparse,
       ]
       #else // os(OpenBSD)
       let userSettableFlags: FileFlags = [
@@ -384,13 +385,13 @@ private struct StatTests {
       #endif
 
       flags.insert(userSettableFlags)
-      try #require(fchflags(fd.rawValue, flags.rawValue) == 0, "\(Errno.current)")
+      try #require(fchflags(fd.rawValue, UInt(flags.rawValue)) == 0, "\(Errno.current)")
 
       stat = try fd.stat()
       #expect(stat.flags == flags)
 
       flags.remove(userSettableFlags)
-      try #require(fchflags(fd.rawValue, flags.rawValue) == 0, "\(Errno.current)")
+      try #require(fchflags(fd.rawValue, UInt(flags.rawValue)) == 0, "\(Errno.current)")
 
       stat = try fd.stat()
       #expect(stat.flags == flags)

